pv211_utils/evaluation_metrics.py

```diff
@@ -0,0 +1,335 @@
+                num_relevant_topk += 1
+        current_rank += 1
+
+    recall = num_relevant_topk / num_relevant
+
+    mr_score_lock.acquire()
+    mr_score.value += recall
+    mr_score_lock.release()
+
+
+def _calc_precision(system: IRSystemBase, judgements: Set, k: int,
+                    mp_score_lock, mp_score: ValueProxy, query: QueryBase) -> None:
+    num_relevant = 0
+    precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+        current_rank += 1
+
+    precision = num_relevant / k
+
+    mp_score_lock.acquire()
+    mp_score.value += precision
+    mp_score_lock.release()
+
+
+def _calc_average_precision(system: IRSystemBase, judgements: Set, k: int,
+                            map_score_lock, map_score: ValueProxy, query: QueryBase) -> None:
+    num_relevant = 0
+    average_precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            average_precision += num_relevant / current_rank
+        current_rank += 1
+
+    average_precision /= num_relevant if num_relevant > 0 else 1
+
+    map_score_lock.acquire()
+    map_score.value += average_precision
+    map_score_lock.release()
+
+
+def _calc_ndcg(system: IRSystemBase, judgements: Set, k: int,
+               ndcg_score_lock, ndcg_score: ValueProxy, query: QueryBase) -> None:
+    num_relevant = 0
+    dcg = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            dcg += 1 / log2(current_rank + 1)
+        current_rank += 1
+
+    # max to avoid division by 0
+    idcg = max(sum([1 / log2(i + 1) for i in range(1, num_relevant + 1)]), 1)
+
+    ndcg_score_lock.acquire()
+    ndcg_score.value += dcg / idcg
+    ndcg_score_lock.release()
+
+
+def _calc_bpref(system: IRSystemBase, judgements: Set, k: int,
+                bpref_score_lock, bpref_score: ValueProxy, query: QueryBase) -> None:
+    num_relevant = 0
+    relevant_doc_ranks = []
+    current_rank = 1
+    bpref = 0.0
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            relevant_doc_ranks.append(current_rank)
+        current_rank += 1
+
+    for i, rank in enumerate(relevant_doc_ranks, start=1):
+        bpref += 1 - min(rank - i, num_relevant) / num_relevant
+
+    bpref_score_lock.acquire()
+    bpref_score.value += bpref / num_relevant
+    bpref_score_lock.release()
+
+
+def mean_average_precision(system: IRSystemBase, queries: OrderedDict,
```

Comment:
This function duplicates `pv211_utils.eval.EvaluationBase.mean_average_precision`. Can you please make `EvaluationBase` use this new implementation instead and remove all code related to the MAP score from `pv211_utils.eval`.

Reply:
Should be done, I also added an option to choose the evaluation depth into the EvaluationBase class and its children.